### PR TITLE
When dumping syncable traits, use a struct for the synced values.

### PR DESCRIPTION
**Pull request by RoosterDragon**
Since most traits have few syncable members, this allows us to avoid allocating an array whose lifetime is only a few ticks long. For traits with more members, we fall back to allocating the array.

Allocating object arrays accounts for 22% of ongoing allocations at game start, this PR reduces it to 2%.

### CODE REVIEW

**Comment by pchote** [comment]
Do you have a histogram or other stats on the numbers here?  I'm wondering whether there is any value with adding additional struct types that can hold more/less elements.

**Comment by RoosterDragon** [comment]
After running the RA shellmap for ~500 ticks, this was how many 'DumpSyncTrait' calls were made with different counts of syncable members:

Count | Occurrences
----|---:
1 | 1164340
2 | 70854
3 | 55080
4 | 23683
5 | 865
6 | 25356
7 | 0
8 | 3470

We could happily just add/remove slots in the struct as we like.

File: OpenRA.Game/Network/SyncReport.cs [hunk 1]
```diff
@@ -258,5 +257,68 @@ public TypeInfo(Type type)
...
 			}
 		}
+
+		/// <summary>
+		/// Holds up to 4 objects directly, or else allocates an array to hold the items. This allows us to record
+		/// trait values for traits with up to 4 sync members inline without having to allocate extra memory.
+		/// </summary>
+		struct Values
+		{
+			static readonly object Sentinal = new object();
```
Comment (pchote) [inline OpenRA.Game/Network/SyncReport.cs]: `Sentinel`?

**Comment by pchote** [comment]
The majority of instances have only one syncable member, so won't this inflate the memory usage?
Can we introduce a fallback to avoid a wrapper and store the value directly if there's only one without introducing a boxing overhead?

**Comment by RoosterDragon** [comment]
We could keep a single entry and check if the entry is an `object[]`, assuming that if it's an array we instead deem it to be an array holding multiple entries. But this still requires allocating arrays for greater than one element.

The goal here is to reduce GC allocations, and trading a bit of overhead for that seems fine to me. 1Mb is enough to store 13000 traits & effects per report. The RA shellmap has about 8000.

If I expand the original table a bit, perhaps that makes my choice of 4 seem less arbitrary. `Bytes` is the size needed by the array which is 4 bytes/element plus an overhead of 8 bytes for the managed object header.

Count | Occurrences | Bytes | Cumulative Allocation Saved %
--|-----------:|-------------:|------:
1 | 1 164 340 | 13 972 080 |  79%
2 |      70 854 |   1 133 664 |  85%
3 |      55 080 |   1 101 600 |  91%
4 |      23 683 |      568 392 |  95%
5 |           865 |        24 220 |  95%
6 |      25 356 |      811 392 |  99%
7 |               0 |                 0 |  99%
8 |        3 470 |      138 800 | 100%

I think I came up with a table like this when I was originally messing with this (it was months ago), and 4 seemed like a sweet spot for savings versus overhead.

File: OpenRA.Game/Network/SyncReport.cs (hunk 1 above)
Comment (RoosterDragon) [reply OpenRA.Game/Network/SyncReport.cs]: Fixed.

**Comment by pchote** [comment]
Ok, that's reasonable so lets stick with the current approach.

**Comment by RoosterDragon** [comment]
(Dear people from the future - you can totally change the number of elements if you like - the table was just how I initially decided to use 4! Please don't feel like it magically must be 4 or anything)

**Review by pchote — APPROVED** [review]

**Comment by penev92** [comment]
👍  but I have a question: when do the allocated arrays get cleaned up? We have an `object`, which is an array, that is a property of a `struct`. I actually don't know who's going to clean that up and when, since we don't explicitly destroy any of this.

**Comment by penev92** [comment]
OK, good enough for me. I trust you know your stuff ;)

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/5300947b8bdabe05f74d84751f76cd9044d7409d...608e187f914996846e40ace4864c22c7e46cf292)

**Pull request merged**